### PR TITLE
feat: Add width / height to ImageTexture

### PR DESCRIPTION
**Pull request by pecoram**
should to fix https://github.com/lightning-js/renderer/issues/34

- without width and height the image width, height are set
- with width, height is set with aspect ratio
- with height, width is set with aspect ratio

### CODE REVIEW

**Review by frank-weindel — CHANGES_REQUESTED** [review]
Thanks for the PR @pecoram! Please discuss with us in #34. I'm not a fan of tying this directly to 0 values and we do need to consider Text in how this functions.

**Comment by frank-weindel** [comment]
Labeling this as a breaking change as it's possible downstream frameworks/apps will behave differently without changes when this feature is implemented.

File: src/core/CoreNode.ts [hunk 1]
```diff
@@ -164,7 +166,34 @@ export class CoreNode extends EventEmitter implements ICoreNode {
     this.props.textureOptions = null;
   }
 
+  autosizeNode(dimensions: Dimensions) {
+    if (this.autosize && dimensions.width && dimensions.height) {
+      if (
+        (isNaN(this.width) && isNaN(this.height)) ||
+        (this.width === 0 && this.height === 0)
```
Comment (frank-weindel) [inline src/core/CoreNode.ts:173]: We shouldn't try to handle aspect ratios. If the `autosize` property is true, it should just set the width/height to the width and height of the texture.

File: src/core/CoreNode.ts [hunk 2]
```diff
@@ -164,7 +166,34 @@ export class CoreNode extends EventEmitter implements ICoreNode {
...
+          this.width = this.height * imageAspectRatio;
+        } else if (isNaN(this.height) || this.height === 0) {
+          this.height = this.width / imageAspectRatio;
+        }
+      }
+    }
+  }
+
   private onTextureLoaded: TextureLoadedEventHandler = (target, dimensions) => {
+    if (target.props.src) {
```
Comment (frank-weindel) [inline src/core/CoreNode.ts:193]: Keep in mind that the `src` property does not exist at all in the Core API Space. It's implementation only exists in the Main API Space, split between MainOnlyNode.ts and ThreadXMainOnlyNode.ts.

https://github.com/lightning-js/renderer/blob/6e9dc53d0dd58a7c949e31d5d4db0770e8b01f5e/src/render-drivers/main/MainOnlyNode.ts#L347C12-L363

https://github.com/lightning-js/renderer/blob/6e9dc53d0dd58a7c949e31d5d4db0770e8b01f5e/src/render-drivers/threadx/ThreadXMainNode.ts#L109-L125

**Review by chiefcll — COMMENTED** [review]
👍

File: src/main-api/INode.ts [hunk 3]
```diff
@@ -98,7 +97,7 @@ export interface INodeWritableProps {
    * TODO: Add support for non-rectangular clipping either automatically or
    * via Render-To-Texture.
    *
-   * @default `false`
+   * @default `null`
```
Comment (frank-weindel) [inline src/main-api/INode.ts:100]: I think this is unintended

File: src/main-api/INode.ts [hunk 4]
```diff
@@ -77,11 +77,10 @@ export interface INodeWritableProps {
    *
    * @remarks
    * The autosize to true let the image node to be resized with the original size of the image.
-   * Combined with one of width and height the other size will be set manteining the aspect ratio
    *
    * @default `false`
    */
-  autosize: boolean;
+  autosize: boolean | null;
```
Comment (frank-weindel) [inline src/main-api/INode.ts:83]: `null` will not work as a value on the ThreadX implementation. For now, just make setting `src` (in both MainOnlyNode and ThreadXMainNode) also set `autosize` to true unconditionally

Comment (frank-weindel) [inline src/main-api/INode.ts:83]: `null` will not work as a value on the ThreadX implementation. For now, just make setting `src` (in both MainOnlyNode and ThreadXMainNode) also set `autosize` to true unconditionally

File: src/core/CoreNode.ts [hunk 5]
```diff
@@ -168,31 +168,13 @@ export class CoreNode extends EventEmitter implements ICoreNode {
 
   autosizeNode(dimensions: Dimensions) {
     if (this.autosize && dimensions.width && dimensions.height) {
```
Comment (frank-weindel) [inline src/core/CoreNode.ts:170]: I don't think we need to condition this on width/height being truthy. If the dimensions are truly somehow 0x0 then autosizing should still occur

File: src/main-api/INode.ts (hunk 4 above)
Comment (frank-weindel) [reply src/main-api/INode.ts:83]: I understand what you're trying to do here. But at the moment there's a limitation in how that can be done. For now we'll need to rely on the order of setting the properties to allow the user to set `autosize` to false if they also set `src`. This is something we can improve on in the future.

**Pull request merged**